internal/directives/git_pusher.go

```diff
@@ -116,18 +120,49 @@ func (g *gitPushPusher) runPromotionStep(
 	}
 	targetBranch := pushOpts.TargetBranch
 	if targetBranch == "" {
-		// If retBranch is still empty, we want to set it to the current branch
+		// If targetBranch is still empty, we want to set it to the current branch
 		// because we will want to return the branch that was pushed to, but we
 		// don't want to mess with the options any further.
 		if targetBranch, err = workTree.CurrentBranch(); err != nil {
 			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
 				fmt.Errorf("error getting current branch: %w", err)
 		}
 	}
-	if err = workTree.Push(pushOpts); err != nil {
+
+	backoff := retry.DefaultBackoff
+	if cfg.MaxAttempts != nil {
+		// Note, the docs for this field say:
+		//
+		//   The remaining number of iterations in which the duration
+		//   parameter may change...
+		//
+		// This is misleading, as it implies that the total number of attempts may
+		// exceed the value of Steps and that Steps only dictates the maximum number
+		// of adjustments to the interval between retries.
+		//
+		// Reading the implementation of retry.DefaultBackoff reveals that Steps
+		// is indeed the maximum number of attempts.
+		backoff.Steps = int(*cfg.MaxAttempts)
+	} else {
+		backoff.Steps = 50
```

Comment:
This default seems pretty high, I think something like `10` would probably be better. Unless you have a specific reason to opt for this value?

Reply:
With the following, and ~20 promos running concurrently and writing to the same branch, I was _occasionally_ able to exhaust attempts:

```
wait.Backoff{
	Duration: 1 * time.Second,
	Factor:   2,
	Steps:    10,
	Cap:      2 * time.Minute,
	Jitter:   0.1,
}
```

Being aware of how extensive some users' monorepos are, I presumed 10 to be not _nearly_ enough.

The effective defaults are now:

```
wait.Backoff{
	Steps:    50,
	Duration: 10 * time.Millisecond,
	Factor:   5.0,
	Jitter:   0.1,
}
```

I think I still like defaulting to a reasonably high number of steps, but what I definitely _didn't_ account for here is that `retry.DefaultBackoff` has no cap -- which makes 50 steps absurd.

What do you think about leaving it 50, but capping it at something like 5s?

Reply:
Alternatively, I could use a mutex to make the pull + rebase + push a critical section of code.

Given that controllers can be sharded, it won't eliminate the need for retries _entirely_, but it _would_ improve the situation considerably.

Reply:
I was using this default backoff because, iirc, you'd suggested it, but I'm rethinking that.

In general, the best way to avoid conflicts will be spacing out the retries of the multiple concurrent promos...

That's got me thinking about a larger starting interval and a larger jitter.

At the same time, increasing the interval by a factor of five at each step seems way too aggressive. Even two seems aggressive.

I'm contemplating something like this:

```
wait.Backoff{
	Steps:    10,
	Duration: time.Second,
	Factor:   1.5,
	Jitter:   0.5,
	Cap:   30 * time.Second,
}
```

I'll try that in the morning to see how it performs.